### PR TITLE
Build: Move Prerequisites from build.ps1 to Invoke-Build script

### PR DESCRIPTION
**Pull request by JustinGrote**
## PR Summary

This is the first in several build process optimizations. 

- Moves all prerequisite checks/logic from build.ps1 to the build script
- Makes the build.ps1 basically a bootstrap for invoke-build only 
- Makes prerequisite assertions idempotent so they can be run on every invocation without a heavy performance penalty
- Refactor the high-level build tasks to be abstracted with their actual implementations as sub-tasks.

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets.
Please mark anything not applicable to this PR `NA`.

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] PR has tests
- [x] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

### CODE REVIEW

**Comment by andschwa** [comment]
I gotta be honest, I've never looked at nor used `build.ps1` 😬

File: .vsts-ci/azure-pipelines-release.yml [hunk 1]
```diff
@@ -4,6 +4,9 @@ variables:
   # Don't download unneeded packages
   - name: DOTNET_SKIP_FIRST_TIME_EXPERIENCE
     value: 'true'
+  - name: DOTNET_NOLOGO
+    value: 'true'
+
```
Comment (andschwa) [inline .vsts-ci/azure-pipelines-release.yml:9]:
```suggestion
  # Suppress welcome message
  - name: DOTNET_NOLOGO
    value: 'true'
```

File: .vsts-ci/azure-pipelines-ci.yml [hunk 2]
```diff
@@ -4,6 +4,8 @@ variables:
   # Don't download unneeded packages
   - name: DOTNET_SKIP_FIRST_TIME_EXPERIENCE
     value: 'true'
+  - name: DOTNET_NOLOGO
+    value: 'true'
```
Comment (andschwa) [inline .vsts-ci/azure-pipelines-ci.yml:8]:
```suggestion
  # Suppress welcome message
  - name: DOTNET_NOLOGO
    value: 'true'
```

File: .vsts-ci/templates/ci-general.yml [hunk 3]
```diff
@@ -55,15 +55,28 @@ steps:
     version: 6.0.x
     performMultiLevelLookup: true
 
+# Some modules like Pester have .NET DLLs that will require a new PowerShell instance, so we run this first prior to the build.
```
Comment (andschwa) [inline .vsts-ci/templates/ci-general.yml:58]: Can you explain a little further? Because we install prerequisites later too.

File: build.ps1
Comment (andschwa) [inline build.ps1:1]: I'm just going to ignore this file since I don't use it 💁

File: requirements.psd1 [hunk 4]
```diff
@@ -0,0 +1,9 @@
+@{
+    Pwsh    = '7.2.0'
+    Node    = '16.14.2'
+    Modules = @(
+        @{ ModuleName = 'Pester'; RequiredVersion = '5.4.1.0' } #For PSES Build. TODO: Remove once we hook into PSES build script
+        @{ ModuleName = 'PSScriptAnalyzer'; RequiredVersion = '1.21.0' } #For PSES Build
+        @{ ModuleName = 'platyPS'; RequiredVersion = '0.14.2' } #For PSES Build
```
Comment (andschwa) [inline requirements.psd1:7]:
```suggestion
        @{ ModuleName = 'Pester'; RequiredVersion = '5.4.1.0' } # For PSES Build. TODO: Remove once we hook into PSES build script
        @{ ModuleName = 'PSScriptAnalyzer'; RequiredVersion = '1.21.0' } # For PSES Build
        @{ ModuleName = 'platyPS'; RequiredVersion = '0.14.2' } # For PSES Build
```

Comment (andschwa) [reply requirements.psd1:7]: But also, shouldn't PSES be installing and controlling these prerequisites?

File: vscode-powershell.build.ps1
Comment (andschwa) [inline vscode-powershell.build.ps1:1]: Ehh I am really hesitant to take these changes as there's only a few of us that build the project on a regular basis anyway, and I really don't want to break CI/CD.

**Comment by andschwa** [comment]
Sounds good. Just remember the release is built _entirely_ via automation in the pipeline...and I've already in the past been bit by bugs like accidentally building (and shipping) debug instead of release, or skipping tests that should've been running. It's just so easy to mess up, especially since the release pipeline depends on assets being produced by the PSES pipeline and handed over to the VS Code pipeline, which has to explicitly use those instead of rebuilding PSES. And that's difficult to test for me in the first place because it's really only testable by going through the release process, and impossible for you to test because that's Microsoft-internal 😓

**Comment by JustinGrote** [comment]
Closing this for now since it probably needs a huge rebase at this point but may revisit some lessons learned into smaller PRs.

**Comment by andyleejordan** [comment]
The build's gotten simpler since then too.

**Pull request closed without merging**